### PR TITLE
Only select used fields of `translation` tables

### PR DESCRIPTION
**Pull request by svenmuennich**
### 1. Why is this change necessary?

Previously anytime a translated field was part of a query a full sub-select of the respective `_translation` table was joined to the main query. A "full sub-select" means a `SELECT` of all columns of the `_translation` table. This can result in a lot of overhead because depending on the storage engine of the database, e.g. `mediumtext` fields like `product_translation.description` require more page reads (i.e. disk access).

In complex, high cardinality queries, especially when they involve e.g. the `product` table, which doubles the number of joined translation tables due to requiring entity inheritance, this overhead can negatively influence the query execution plans. In fact, the same query (filtering orders based on their line items and respective products) that took a reasonable time in Shopware 6.5 ballooned to several hundred seconds with Shopware 6.6 due to a correction in filter evaluation, which leads to more joins of sub-selects in general.

### 2. What does this change do, exactly?

* This PR reduces the selected fields of joined `_translation` tables to those that are actually needed for filtering or resolving `TranslatedField`s. It does not affect queries of `_translation` tables that are needed for loading translation associations.
* This PR renames `fallback_` join aliases in the generated queries to use `override_` aliases, because their values in fact _override_ the default translations, if available. This is just cosmetics, but IMHO makes the code (and generated queries) easier to understand.
* This PR adds some more escaping to the queries that join `_translation` tables.
* **This PR currently does not add any new tests.** ~I would like to get feedback first on the general approach as well as what kind of tests (unit, integration) are expected for these changes.~ All tests added in https://github.com/shopware/shopware/pull/3839 should still pass.

### 3. Describe each step to reproduce the issue or behaviour.

For simply seeing the difference in the generated SQL queries turn on your database's query log and send the following request to the admin API:
```
POST /api/search/product

{
    "filter": [
        {
            "type": "contains",
            "field": "name",
            "value": "foo"
        }
    ]
}
```

<details>
    <summary>Generated SQL (old)</summary>

    SELECT
        `product`.`id`,
        `product`.`product_number`,
        `product`.`auto_increment`
    FROM `product`
    LEFT JOIN `product` `product.parent`
        ON `product`.`parent_id` = `product.parent`.`id`
        AND `product`.`parent_version_id` = `product.parent`.`version_id`
    LEFT JOIN (
        SELECT
            `product.translation`.product_id,
            `product.translation`.product_version_id,
            `product.translation`.meta_description AS `product.translation.metaDescription`,
            `product.translation`.name AS `product.translation.name`,
            `product.translation`.keywords AS `product.translation.keywords`,
            `product.translation`.description AS `product.translation.description`,
            `product.translation`.meta_title AS `product.translation.metaTitle`,
            `product.translation`.pack_unit AS `product.translation.packUnit`,
            `product.translation`.pack_unit_plural AS `product.translation.packUnitPlural`,
            `product.translation`.custom_search_keywords AS `product.translation.customSearchKeywords`,
            `product.translation`.slot_config AS `product.translation.slotConfig`,
            `product.translation`.custom_fields AS `product.translation.customFields`,
            `product.translation`.created_at AS `product.translation.createdAt`,
            `product.translation`.updated_at AS `product.translation.updatedAt`,
            `product.translation`.product_id AS `product.translation.productId`,
            `product.translation`.language_id AS `product.translation.languageId`,
            `product.translation`.product_version_id AS `product.translation.productVersionId`,
            `product.translation.fallback_1`.meta_description AS `product.translation.fallback_1.metaDescription`,
            `product.translation.fallback_1`.name AS `product.translation.fallback_1.name`,
            `product.translation.fallback_1`.keywords AS `product.translation.fallback_1.keywords`,
            `product.translation.fallback_1`.description AS `product.translation.fallback_1.description`,
            `product.translation.fallback_1`.meta_title AS `product.translation.fallback_1.metaTitle`,
            `product.translation.fallback_1`.pack_unit AS `product.translation.fallback_1.packUnit`,
            `product.translation.fallback_1`.pack_unit_plural AS `product.translation.fallback_1.packUnitPlural`,
            `product.translation.fallback_1`.custom_search_keywords AS `product.translation.fallback_1.customSearchKeywords`,
            `product.translation.fallback_1`.slot_config AS `product.translation.fallback_1.slotConfig`,
            `product.translation.fallback_1`.custom_fields AS `product.translation.fallback_1.customFields`,
            `product.translation.fallback_1`.created_at AS `product.translation.fallback_1.createdAt`,
            `product.translation.fallback_1`.updated_at AS `product.translation.fallback_1.updatedAt`,
            `product.translation.fallback_1`.product_id AS `product.translation.fallback_1.productId`,
            `product.translation.fallback_1`.language_id AS `product.translation.fallback_1.languageId`,
            `product.translation.fallback_1`.product_version_id AS `product.translation.fallback_1.productVersionId`
        FROM `product_translation` `product.translation`
        LEFT JOIN `product_translation` `product.translation.fallback_1`
            ON `product.translation`.`product_id` = `product.translation.fallback_1`.`product_id`
            AND `product.translation.fallback_1`.language_id = :languageId
            AND `product.translation`.product_version_id = `product.translation.fallback_1`.product_version_id
        WHERE
            `product.translation`.language_id = :languageId
    ) `product.product_translation`
        ON `product.product_translation`.`product_id` = `product`.`id`
        AND `product.product_translation`.product_version_id = `product`.version_id
    LEFT JOIN (
        SELECT
            `product.parent.translation`.product_id,
            `product.parent.translation`.product_version_id,
            `product.parent.translation`.meta_description AS `product.parent.translation.metaDescription`,
            `product.parent.translation`.name AS `product.parent.translation.name`,
            `product.parent.translation`.keywords AS `product.parent.translation.keywords`,
            `product.parent.translation`.description AS `product.parent.translation.description`,
            `product.parent.translation`.meta_title AS `product.parent.translation.metaTitle`,
            `product.parent.translation`.pack_unit AS `product.parent.translation.packUnit`,
            `product.parent.translation`.pack_unit_plural AS `product.parent.translation.packUnitPlural`,
            `product.parent.translation`.custom_search_keywords AS `product.parent.translation.customSearchKeywords`,
            `product.parent.translation`.slot_config AS `product.parent.translation.slotConfig`,
            `product.parent.translation`.custom_fields AS `product.parent.translation.customFields`,
            `product.parent.translation`.created_at AS `product.parent.translation.createdAt`,
            `product.parent.translation`.updated_at AS `product.parent.translation.updatedAt`,
            `product.parent.translation`.product_id AS `product.parent.translation.productId`,
            `product.parent.translation`.language_id AS `product.parent.translation.languageId`,
            `product.parent.translation`.product_version_id AS `product.parent.translation.productVersionId`,
            `product.parent.translation.fallback_1`.meta_description AS `product.parent.translation.fallback_1.metaDescription`,
            `product.parent.translation.fallback_1`.name AS `product.parent.translation.fallback_1.name`,
            `product.parent.translation.fallback_1`.keywords AS `product.parent.translation.fallback_1.keywords`,
            `product.parent.translation.fallback_1`.description AS `product.parent.translation.fallback_1.description`,
            `product.parent.translation.fallback_1`.meta_title AS `product.parent.translation.fallback_1.metaTitle`,
            `product.parent.translation.fallback_1`.pack_unit AS `product.parent.translation.fallback_1.packUnit`,
            `product.parent.translation.fallback_1`.pack_unit_plural AS `product.parent.translation.fallback_1.packUnitPlural`,
            `product.parent.translation.fallback_1`.custom_search_keywords AS `product.parent.translation.fallback_1.customSearchKeywords`,
            `product.parent.translation.fallback_1`.slot_config AS `product.parent.translation.fallback_1.slotConfig`,
            `product.parent.translation.fallback_1`.custom_fields AS `product.parent.translation.fallback_1.customFields`,
            `product.parent.translation.fallback_1`.created_at AS `product.parent.translation.fallback_1.createdAt`,
            `product.parent.translation.fallback_1`.updated_at AS `product.parent.translation.fallback_1.updatedAt`,
            `product.parent.translation.fallback_1`.product_id AS `product.parent.translation.fallback_1.productId`,
            `product.parent.translation.fallback_1`.language_id AS `product.parent.translation.fallback_1.languageId`,
            `product.parent.translation.fallback_1`.product_version_id AS `product.parent.translation.fallback_1.productVersionId`
        FROM `product_translation` `product.parent.translation`
        LEFT JOIN `product_translation` `product.parent.translation.fallback_1`
            ON `product.parent.translation`.`product_id` = `product.parent.translation.fallback_1`.`product_id`
            AND `product.parent.translation.fallback_1`.language_id = :languageId
            AND `product.parent.translation`.product_version_id = `product.parent.translation.fallback_1`.product_version_id
        WHERE
            `product.parent.translation`.language_id = :languageId
    ) `product.product_translation.parent`
        ON `product.product_translation.parent`.`product_id` = `product.parent`.`id`
        AND `product.product_translation.parent`.product_version_id = `product.parent`.version_id
    WHERE
        (`product`.`version_id` = :versionId)
        AND ((
            COALESCE(
                `product.translation.fallback_1.name`,
                `product.parent.translation.fallback_1.name`,
                `product.translation.name`,
                `product.parent.translation.name`
            ) LIKE :query
        ))
    LIMIT 500
</details>
<details>
    <summary>Generated SQL (new)</summary>

    SELECT
        `product`.`id`,
        `product`.`product_number`,
        `product`.`auto_increment`
    FROM `product`
    LEFT JOIN `product` `product.parent`
        ON `product`.`parent_id` = `product.parent`.`id`
        AND `product`.`parent_version_id` = `product.parent`.`version_id`
    LEFT JOIN (
        SELECT
            `product.translation`.`product_id`,
            `product.translation`.`product_version_id`,
            `product.translation`.`name` AS `product.translation.name`,
            `product.translation.override_1`.`name` AS `product.translation.override_1.name`
        FROM `product_translation` `product.translation`
        LEFT JOIN `product_translation` `product.translation.override_1`
            ON `product.translation`.`product_id` = `product.translation.override_1`.`product_id`
            AND `product.translation.override_1`.`language_id` = :languageId
            AND `product.translation`.`product_version_id` = `product.translation.override_1`.`product_version_id`
        WHERE
            `product.translation`.`language_id` = :languageId
    ) `product.product_translation`
        ON `product.product_translation`.`product_id` = `product`.`id`
        AND `product.product_translation`.`product_version_id` = `product`.`version_id`
    LEFT JOIN (
        SELECT
            `product.parent.translation`.`product_id`,
            `product.parent.translation`.`product_version_id`,
            `product.parent.translation`.`name` AS `product.parent.translation.name`,
            `product.parent.translation.override_1`.`name` AS `product.parent.translation.override_1.name`
        FROM `product_translation` `product.parent.translation`
        LEFT JOIN `product_translation` `product.parent.translation.override_1`
            ON `product.parent.translation`.`product_id` = `product.parent.translation.override_1`.`product_id`
            AND `product.parent.translation.override_1`.`language_id` = :languageId
            AND `product.parent.translation`.`product_version_id` = `product.parent.translation.override_1`.`product_version_id`
        WHERE
            `product.parent.translation`.`language_id` = :languageId
    ) `product.parent.product_translation`
        ON `product.parent.product_translation`.`product_id` = `product.parent`.`id`
        AND `product.parent.product_translation`.`product_version_id` = `product.parent`.`version_id`
    WHERE
        (`product`.`version_id` = :versionId)
        AND ((
            COALESCE(
                `product.translation.override_1.name`,
                `product.parent.translation.override_1.name`,
                `product.translation.name`,
                `product.parent.translation.name`
            ) LIKE :query
        ))
    LIMIT 500
</details>

That simple example will usually not show a difference in performance (nor query execution plan). The following request is taken from one of our applications and will drastically improve in execution time with the fix (we noticed ~60x speed up in production environments; 215s (old) vs. 3,5s (new); 20k orders, 80k order line items, 1,5k products, MySQL v8.0.34):
```
POST /api/search/order

{
    "filter": [
        {
            "type": "multi",
            "operator": "and",
            "queries": [
                {
                    "type": "equals",
                    "field": "lineItems.type",
                    "value": "product"
                },
                {
                    "type": "not",
                    "operator": "and",
                    "queries": [
                        {
                            "type": "equals",
                            "field": "lineItems.product.id",
                            "value": null
                        }
                    ]
                },
                {
                    "type": "equalsAny",
                    "field": "stateMachineState.technicalName",
                    "value": [
                        "open",
                        "in_progress"
                    ]
                },
                {
                    "type": "equalsAny",
                    "field": "transactions.stateMachineState.technicalName",
                    "value": [
                        "paid",
                        "refunded_partially"
                    ]
                },
                {
                    "type": "multi",
                    "operator": "or",
                    "queries": [
                        {
                            "type": "prefix",
                            "field": "orderNumber",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.firstName",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.lastName",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.company",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.department",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.street",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.additionalAddressLine1",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.additionalAddressLine2",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.zipcode",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "value": "Foo",
                            "field": "billingAddress.city"
                        },
                        {
                            "type": "contains",
                            "field": "billingAddress.country.name",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "customerComment",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.firstName",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.lastName",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.company",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.department",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.street",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.additionalAddressLine1",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.additionalAddressLine2",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.zipcode",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.city",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "deliveries.shippingOrderAddress.country.name",
                            "value": "Foo"
                        },
                        {
                            "field": "deliveries.shippingMethod.name",
                            "type": "contains",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "lineItems.product.name",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "lineItems.product.productNumber",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "lineItems.product.ean",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "lineItems.product.manufacturer.name",
                            "value": "Foo"
                        },
                        {
                            "type": "contains",
                            "field": "lineItems.product.manufacturerNumber",
                            "value": "Foo"
                        }
                    ]
                }
            ]
        }
    ]
}
```

### 4. Please link to the relevant issues (if any).

n/a

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: src/Core/Framework/DataAbstractionLayer/Dbal/EntityDefinitionQueryHelper.php [hunk 1]
```diff
@@ -449,50 +448,37 @@ public function addTranslationSelect(string $root, EntityDefinition $definition,
...
-        $fields = $translationDefinition->getFields();
+        $fields = $translationDefinition->getFields()->filter(
+            fn (Field $field) => $field instanceof StorageAware
+                && $definition->getFields()->get($field->getPropertyName()) instanceof TranslatedField,
+        );
         if (!empty($partial)) {
-            $fields = $translationDefinition->getFields()->filter(fn (Field $field) => $field->is(PrimaryKey::class)
-                || isset($partial[$field->getPropertyName()])
-                || $field instanceof FkField);
+            $fields = $fields->filter(fn (Field $field) => isset($partial[$field->getPropertyName()]));
```
Comment (shyim) [inline src/Core/Framework/DataAbstractionLayer/Dbal/EntityDefinitionQueryHelper.php:456]: why did you removed the primary key and fk field here. seems not related to your changes?

Comment (svenmuennich) [reply src/Core/Framework/DataAbstractionLayer/Dbal/EntityDefinitionQueryHelper.php:456]: Because they were never read.

They just used to end up in the outer-most `SELECT` when fetching the entities for hydration but were never accessed. Just like `TranslationFieldResolver` this method used to select more fields than necessary, so I removed them.

Here are example queries I just dumped using the following criteria (`fields` to apply the `$partial` filter):
```
{
    "fields": [
        "name"
    ],
    "filter": [
        {
            "type": "contains",
            "field": "name",
            "value": "a"
        }
    ]
}
```
Old query:
```sql
SELECT
    `product`.`id` AS `product.id`,
    `product`.`version_id` AS `product.versionId`,
    # [START] Selects added by EntityDefinitionQueryHelper::addTranslationSelect()
    `product.translation.fallback_1.name`,
    `product.parent.translation.fallback_1.name`,
    `product.translation.name`,
    `product.parent.translation.name`,
    COALESCE(
        `product.translation.fallback_1.name`,
        `product.parent.translation.fallback_1.name`,
        `product.translation.name`,
        `product.parent.translation.name`,
        `product.translation.name`
    ) AS `product.name`,
    `product.translation.fallback_1.productId`,
    `product.parent.translation.fallback_1.productId`,
    `product.translation.productId`,
    `product.parent.translation.productId`,
    `product.translation.fallback_1.languageId`,
    `product.parent.translation.fallback_1.languageId`,
    `product.translation.languageId`,
    `product.parent.translation.languageId`,
    `product.translation.fallback_1.productVersionId`,
    `product.parent.translation.fallback_1.productVersionId`,
    `product.translation.productVersionId`,
    `product.parent.translation.productVersionId`
    # [START] Selects added by EntityDefinitionQueryHelper::addTranslationSelect()
FROM `product`
LEFT JOIN `product` `product.parent`
    ON `product`.`parent_id` = `product.parent`.`id`
    AND `product`.`parent_version_id` = `product.parent`.`version_id`
LEFT JOIN (
    SELECT
        `product.translation`.product_id,
        `product.translation`.product_version_id,
        `product.translation`.meta_description AS `product.translation.metaDescription`,
        `product.translation`.name AS `product.translation.name`,
        `product.translation`.keywords AS `product.translation.keywords`,
        `product.translation`.description AS `product.translation.description`,
        `product.translation`.meta_title AS `product.translation.metaTitle`,
        `product.translation`.pack_unit AS `product.translation.packUnit`,
        `product.translation`.pack_unit_plural AS `product.translation.packUnitPlural`,
        `product.translation`.custom_search_keywords AS `product.translation.customSearchKeywords`,
        `product.translation`.slot_config AS `product.translation.slotConfig`,
        `product.translation`.custom_fields AS `product.translation.customFields`,
        `product.translation`.created_at AS `product.translation.createdAt`,
        `product.translation`.updated_at AS `product.translation.updatedAt`,
        `product.translation`.product_id AS `product.translation.productId`,
        `product.translation`.language_id AS `product.translation.languageId`,
        `product.translation`.product_version_id AS `product.translation.productVersionId`,
        `product.translation.fallback_1`.meta_description AS `product.translation.fallback_1.metaDescription`,
        `product.translation.fallback_1`.name AS `product.translation.fallback_1.name`,
        `product.translation.fallback_1`.keywords AS `product.translation.fallback_1.keywords`,
        `product.translation.fallback_1`.description AS `product.translation.fallback_1.description`,
        `product.translation.fallback_1`.meta_title AS `product.translation.fallback_1.metaTitle`,
        `product.translation.fallback_1`.pack_unit AS `product.translation.fallback_1.packUnit`,
        `product.translation.fallback_1`.pack_unit_plural AS `product.translation.fallback_1.packUnitPlural`,
        `product.translation.fallback_1`.custom_search_keywords AS `product.translation.fallback_1.customSearchKeywords`,
        `product.translation.fallback_1`.slot_config AS `product.translation.fallback_1.slotConfig`,
        `product.translation.fallback_1`.custom_fields AS `product.translation.fallback_1.customFields`,
        `product.translation.fallback_1`.created_at AS `product.translation.fallback_1.createdAt`,
        `product.translation.fallback_1`.updated_at AS `product.translation.fallback_1.updatedAt`,
        `product.translation.fallback_1`.product_id AS `product.translation.fallback_1.productId`,
        `product.translation.fallback_1`.language_id AS `product.translation.fallback_1.languageId`,
        `product.translation.fallback_1`.product_version_id AS `product.translation.fallback_1.productVersionId`
    FROM `product_translation` `product.translation`
    LEFT JOIN `product_translation` `product.translation.fallback_1`
        ON `product.translation`.`product_id` = `product.translation.fallback_1`.`product_id`
        AND `product.translation.fallback_1`.language_id = :languageId
        AND `product.translation`.product_version_id = `product.translation.fallback_1`.product_version_id
    WHERE
        `product.translation`.language_id = :languageId
) `product.product_translation`
    ON `product.product_translation`.`product_id` = `product`.`id`
    AND `product.product_translation`.product_version_id = `product`.version_id
LEFT JOIN (
    SELECT
        `product.parent.translation`.product_id,
        `product.parent.translation`.product_version_id,
        `product.parent.translation`.meta_description AS `product.parent.translation.metaDescription`,
        `product.parent.translation`.name AS `product.parent.translation.name`,
        `product.parent.translation`.keywords AS `product.parent.translation.keywords`,
        `product.parent.translation`.description AS `product.parent.translation.description`,
        `product.parent.translation`.meta_title AS `product.parent.translation.metaTitle`,
        `product.parent.translation`.pack_unit AS `product.parent.translation.packUnit`,
        `product.parent.translation`.pack_unit_plural AS `product.parent.translation.packUnitPlural`,
        `product.parent.translation`.custom_search_keywords AS `product.parent.translation.customSearchKeywords`,
        `product.parent.translation`.slot_config AS `product.parent.translation.slotConfig`,
        `product.parent.translation`.custom_fields AS `product.parent.translation.customFields`,
        `product.parent.translation`.created_at AS `product.parent.translation.createdAt`,
        `product.parent.translation`.updated_at AS `product.parent.translation.updatedAt`,
        `product.parent.translation`.product_id AS `product.parent.translation.productId`,
        `product.parent.translation`.language_id AS `product.parent.translation.languageId`,
        `product.parent.translation`.product_version_id AS `product.parent.translation.productVersionId`,
        `product.parent.translation.fallback_1`.meta_description AS `product.parent.translation.fallback_1.metaDescription`,
        `product.parent.translation.fallback_1`.name AS `product.parent.translation.fallback_1.name`,
        `product.parent.translation.fallback_1`.keywords AS `product.parent.translation.fallback_1.keywords`,
        `product.parent.translation.fallback_1`.description AS `product.parent.translation.fallback_1.description`,
        `product.parent.translation.fallback_1`.meta_title AS `product.parent.translation.fallback_1.metaTitle`,
        `product.parent.translation.fallback_1`.pack_unit AS `product.parent.translation.fallback_1.packUnit`,
        `product.parent.translation.fallback_1`.pack_unit_plural AS `product.parent.translation.fallback_1.packUnitPlural`,
        `product.parent.translation.fallback_1`.custom_search_keywords AS `product.parent.translation.fallback_1.customSearchKeywords`,
        `product.parent.translation.fallback_1`.slot_config AS `product.parent.translation.fallback_1.slotConfig`,
        `product.parent.translation.fallback_1`.custom_fields AS `product.parent.translation.fallback_1.customFields`,
        `product.parent.translation.fallback_1`.created_at AS `product.parent.translation.fallback_1.createdAt`,
        `product.parent.translation.fallback_1`.updated_at AS `product.parent.translation.fallback_1.updatedAt`,
        `product.parent.translation.fallback_1`.product_id AS `product.parent.translation.fallback_1.productId`,
        `product.parent.translation.fallback_1`.language_id AS `product.parent.translation.fallback_1.languageId`,
        `product.parent.translation.fallback_1`.product_version_id AS `product.parent.translation.fallback_1.productVersionId`
    FROM `product_translation` `product.parent.translation`
    LEFT JOIN `product_translation` `product.parent.translation.fallback_1`
        ON `product.parent.translation`.`product_id` = `product.parent.translation.fallback_1`.`product_id`
        AND `product.parent.translation.fallback_1`.language_id = :languageId
        AND `product.parent.translation`.product_version_id = `product.parent.translation.fallback_1`.product_version_id
    WHERE
        `product.parent.translation`.language_id = :languageId
) `product.product_translation.parent`
    ON `product.product_translation.parent`.`product_id` = `product.parent`.`id`
    AND `product.product_translation.parent`.product_version_id = `product.parent`.version_id
WHERE
    `product`.`version_id` = :versionId
```

New query:
```sql
SELECT
    `product`.`id` AS `product.id`,
    `product`.`version_id` AS `product.versionId`,
    # [START] Selects added by EntityDefinitionQueryHelper::addTranslationSelect()
    `product.translation.override_1.name`,
    `product.parent.translation.override_1.name`,
    `product.translation.name`,
    `product.parent.translation.name`,
    COALESCE(
        `product.translation.override_1.name`,
        `product.parent.translation.override_1.name`,
        `product.translation.name`,
        `product.parent.translation.name`
    ) AS `product.name`
    # [END] Selects added by EntityDefinitionQueryHelper::addTranslationSelect()
FROM `product`
LEFT JOIN `product` `product.parent`
    ON `product`.`parent_id` = `product.parent`.`id`
    AND `product`.`parent_version_id` = `product.parent`.`version_id`
LEFT JOIN (
    SELECT
        `product.translation`.`product_id`,
        `product.translation`.`product_version_id`,
        `product.translation`.`name` AS `product.translation.name`,
        `product.translation.override_1`.`name` AS `product.translation.override_1.name`
    FROM `product_translation` `product.translation`
    LEFT JOIN `product_translation` `product.translation.override_1`
        ON `product.translation`.`product_id` = `product.translation.override_1`.`product_id`
        AND `product.translation.override_1`.`language_id` = :languageId
        AND `product.translation`.`product_version_id` = `product.translation.override_1`.`product_version_id`
    WHERE
        `product.translation`.`language_id` = :languageId
) `product.product_translation`
    ON `product.product_translation`.`product_id` = `product`.`id`
    AND `product.product_translation`.`product_version_id` = `product`.`version_id`
LEFT JOIN (
    SELECT
        `product.parent.translation`.`product_id`,
        `product.parent.translation`.`product_version_id`,
        `product.parent.translation`.`name` AS `product.parent.translation.name`,
        `product.parent.translation.override_1`.`name` AS `product.parent.translation.override_1.name`
    FROM `product_translation` `product.parent.translation`
    LEFT JOIN `product_translation` `product.parent.translation.override_1`
        ON `product.parent.translation`.`product_id` = `product.parent.translation.override_1`.`product_id`
        AND `product.parent.translation.override_1`.`language_id` = :languageId
        AND `product.parent.translation`.`product_version_id` = `product.parent.translation.override_1`.`product_version_id`
    WHERE
        `product.parent.translation`.`language_id` = :languageId
) `product.parent.product_translation`
    ON `product.parent.product_translation`.`product_id` = `product.parent`.`id`
    AND `product.parent.product_translation`.`product_version_id` = `product.parent`.`version_id`
WHERE
    `product`.`version_id` = :versionId
```

**Comment by AydinHassan** [comment]
@svenmuennich Thanks for the PR, general approach looks fine. We are slightly worried that some breaks might occur so we would appreciate if you tackle the tests and then we can do another round of reviews, wdyt?

**Comment by svenmuennich** [comment]
>would appreciate if you tackle the tests and then we can do another round of reviews

Sure thing. What kind of tests would you like me to add? Are integration tests covering the changed behavior sufficient?

**Comment by AydinHassan** [comment]
@svenmuennich yep integration tests would be useful here.

**Comment by OliverSkroblin** [comment]
Hi, 

LGTM  👍 

It would be good when you test the following scenarios:
- Filter with multiple fields in different levels of a Criteria
- Filter with multiple languages and multiple fields
- Full text search via `term` 
- Full text search `criteria.queries` 
- Mixed searches with filters and queries

**Comment by svenmuennich** [comment]
I added the integration tests in a separate PR (https://github.com/shopware/shopware/pull/3839) to ensure that they cover the current behavior. I also rebased this PR accordingly.

**Comment by AydinHassan** [comment]
@svenmuennich could you take a look at the phpstan failures please?

**Comment by svenmuennich** [comment]
Linting looks good now ...

**Comment by AydinHassan** [comment]
Sorry for the delay @svenmuennich - I was on holiday then battling with pipeline issues :D - Thanks for the improvements, it's merged now!
